src/sentry/ratelimits/leaky_bucket.py

```diff
@@ -0,0 +1,213 @@
+                print(f"Rate limited, needs to wait for {e.wait_time} seconds")
+
+            ```
+
+            - checking info even if not limited:
+            ```
+            try:
+                with limiter.context(key) as limiter_info:
+                    value = do_something()
+            except LeakyBucketRateLimiter.LimitExceeded as e:
+                limiter_info = e.info
+            ```z
+        """
+
+        @contextmanager
+        def limit_context() -> Generator[LeakyBucketLimitInfo, None, None]:
+            info = self.use_and_get_info(key)
+            if info.wait_time:
+                raise self.LimitExceeded(info)
+
+            yield info
+
+        return limit_context()
+
+    def __call__(
+        self,
+        key: str,
+        limited_handler: Callable[[LeakyBucketLimitInfo, dict[str, Any]], Any] | None = None,
+        raise_exception: bool = False,
+    ) -> Callable[[Any], Any]:
+        """
+        Decorator to limit the rate of requests on a given key
+
+        limited_handler: a callback function that will be called when the rate limit is exceeded,
+                            it will receive the LeakyBucketLimitInfo instance as an argument,
+                            as well as original args and kwargs in a dictionary
+                        return value will be returned as the result of the decorated function
+
+        raise_exception: if True, LeakyBucketRateLimiter.LimitExceeded will be raised when the rate limit is exceeded
+
+
+        if both limited_callback and raise_exception are provided, limited_callback will take precedence
+        if neither limited_callback nor raise_exception is provided, the decorated function will silently
+        be ignored when the rate limit is exceeded
+
+
+
+        usage:
+
+            - basic usage:
+            ```
+            limiter = LeakyBucketRateLimiter(burst_limit=10, drip_rate=1)
+
+            @limiter(key)
+            def my_function():
+                do_something()
+            ```
+
+            - raising an exception when limited:
+            ```
+            try:
+                @limiter(key, raise_exception=True)
+                def my_function():
+                    do_something()
+            except LeakyBucketRateLimiter.LimitExceeded as e:
+                print("Rate limited, needs to wait for {e.wait_time} seconds")
+            ```
+
+            - providing a handler function:
+            ```
+            def my_limited_handler(info, context):
+                print(f"Rate limited, needs to wait for {info.wait_time} seconds")
+                print(f"Original args: {context['args']}")
+                print(f"Original kwargs: {context['kwargs']}")
+                return None
+
+            @limiter(key, limited_handler=my_limited_handler)
+            def my_function():
+                rv = do_something()
+            ```
+
+        """
+
+        def decorator(func):
+            def wrapper(*args, **kwargs):
```

Comment:
Is there a reason why we are not using `functools.wraps` decorator here?

Reply:
that's a good point, might be useful if we generate docs or such, I'll add it.